enhancements/image-references.md

```diff
@@ -0,0 +1,228 @@
+  which any individual build system will resolve image references
+- upstream operator manifests must not be required to know how common names or
+  references change when built in a downstream pipeline
+- no prescriptive statement is made about the specific format or contents of the
+  bundle image layer; any will be transparently supported as an output
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an author of a manifest, I would like to check in manifests to my upstream
+repository that are self-consistent, valid and make no assumptions about the
+build system that will eventually create a bundle image with them.
+
+#### Story 2
+
+As an author of an operand, I would like to create a bundle locally in order
+to test my operator end-to-end on a cluster of my choosing without having to
+edit the core configuration for the operator.
+
+#### Story 3
+
+As an author of a build system, I would like to operate with tooling that allows
+me to clearly define the source of truth for image digests in order to keep the
+build-system-specific configuration to an absolute minimum.
+
+#### Story 4
+
+As an engineer involved in publishing an optional operator, I would like to
+configure semantically equivalent image pull specifications once, in order to not
+need to configure each build system independently.
+
+### Implementation Details/Notes/Constraints
+
+The core problems that must be solved in the implementation of this proposal have
+already been handled in the workflow used in `oc adm release new`. When implementing
+improvements to the `operator-sdk generate bundle` process we will simply need to
+create a shared library for the two tools to use. While the shape of the output
+is slightly different and some of the semantics about how the output should be 
+formatted are dissmilar, the core image reference rewriting is identical and the
+process of building a `FROM scratch` image layer is also identical.
+
+Today, some prior art exists in the OSBS workflow for building operator bundles.
+As we improve the Operator SDK tooling to create a straghtforward process for
+creating bundle images, we must make sure a seamless migration is possible.
+
+### Risks and Mitigations
+
+It will be critical that the design be vetted by all of the concerned parties, from
+operator manifest authors to CI system authors and productized pipeline authors to
+ensure that the UX is appropriate in all cases. Furthermore, the largest risk in the
+implementation here is not prioritizing a clean migration pathway for all current
+users who create bundle images, which would lead to further fragmentation of the
+ecosystem, which is directly opposed to the goal of this enhancement.
+
+## Design Details
+
+The definition of a minimally-viable operator bundle image will be changed to
+ensure that all image references in the contained manifests have been resolved 
+to a digest and had the pull specifications rewritten to refer to those digests.
```

Comment:
An image reference that needs pinning can appear in multiple locations in the CSV file:

1. RelatedImage: `spec.relatedImages[].image`
2. Annotation: any of:
   1. `metadata.annotations.containerImage`
   2. `metadata.annotations[]` for any given Object where the annotation value contains one
      or more image pull specification.
3. Container: `spec.install.spec.deployments[].spec.template.spec.containers[].image`
4. InitContainer: `spec.install.spec.deployments[].spec.template.spec.initContainers[].image`
5. RelatedImageEnv: `spec.install.spec.deployments[].spec.template.spec.containers[].env[].value`
   and `spec.install.spec.deployments[].spec.template.spec.initContainers[].env[].value` where the
   `name` of the corresponding env Object is prefixed by `RELATED_IMAGE_`.

This is the currently supported locations for IIB, OSBS, and CVP.

In order to replace an image reference with its pinned form, all of those locations must be searched. 
I don't think there's a way to do the replacements without being aware of those locations. A simple `sed-like` approach will definitely not cover all use cases. `2.ii` is particularly difficult.

I bring up this point because from an implementation POV the process in this proposal **must** be aware of each one of those locations. Otherwise, it cannot replace image references reliably.

It's desirable to have the list of valid image reference locations in a centralized location, currently we have at least 3.

This proposal **must** include a mechanism which can be used to identify which image references are in the CSV.

Reply:
If I understand the proposal correctly, you could get away with not knowing about those locations because we're forcing authors to be explicit about image references. IMO, we're moving complexity from the tooling to the authors which seems like a step backwards. For instance, an author now needs to provide the same placeholder pull spec in `RELATED_IMAGE_FOO` environment variable, again in `spec.relatedImages`, and one more time in image-references.

Reply:
The explicit ask from authors that wish to opt into image pinning pays large dividends by making a simple, transparent and interoperable system. The fact that you just pasted a long list of different ways for someone to explicitly or implicitly opt into it today, and the fact that the resolution process would be duplicated across build pipelines speaks to that. This is an active and explicit decision.

Reply:
> For instance, an author now needs to provide the same placeholder pull spec in RELATED_IMAGE_FOO environment variable, again in spec.relatedImages, and one more time in image-references.

Wheras today ... they need to provide what is effectively a placeholder that happens to also be an image pull spec. I don't see the functional difference here.

Reply:
Consider this example that the build systems can handle today:

```yaml
# csv file
metadata:
  annotations:
    containerImage: registry.example.com/image:tag
```

Once it's processed via the build system, it's transformed into something like:

```yaml
# csv file
metadata:
  annotations:
    containerImage: registry.example.com/image@sha256:123...
spec:
  relatedImages:
    - name: some-name
      value: registry.example.com/image@sha256:123...
```

The authors only have to define the pull spec once.

With this proposal, the author would have to write this:
```yaml
# csv file
metadata:
  annotations:
    containerImage: registry.example.com/image:tag
spec:
  relatedImages:
    - containerImage: registry.example.com/image:tag

# image-references
imageReferences:
- name: common-name
  placeholder: registry.example.com/image:tag
```

Authors have to duplicate the placeholder pull spec at least three times for every unique pull spec across 2 different files. Additionally, authors must now have some sort of static analysis to make sure there are no typos in those pull specs. Otherwise, they may not be replaced as desired, which may be easy to miss if there are 10 different pull spec in the CSV file.

As I mentioned earlier, this is moving the complexity from the build system to the authors.

I do like the idea of removing the current "magic" of finding pull specs in all sorts of places in the CSV file. What if instead of placeholder pull specs, we used an actual templating language? The author could write the following:

```yaml
# csv file
metadata:
  annotations:
    containerImage: {{ COMMON_NAME }}
spec:
  relatedImages:
    - name: common-name
      value: {{ COMMON_NAME }}

# image-references
imageReferences:
- name: COMMON_NAME
  # default value if image-replacements is not provided.
  default: registry.example.com/image:tag
```

We could even go a step further and provide an option to automatically populate `spec.relatedImages` from image-references.

Reply:
You are inventing the requirement for authors to replcate it three times. The proposal as written does not require it. We ask authors to identify their _common names_ for images and the opaque identifier that this common name is represented as. This is all they need to do:

```yaml
# image-references
imageReferences:
- name: common-name
  placeholder: registry.example.com/image:tag
```

You are pushing too much in the wrong direction. Please understand that this explicit mapping between an identifier and the placeholder is critical for interoperability between build systems. This allows the CSV to remain entirely static. If we tie the identifier to the registry/org/name/tag that the CSV happens to use, it is implicit and confusing when you add into the picture:

 - registry renames that occur differently between build systems
 - organization changes that apply to separate registries
 - _name_ changes (for instance, the Bugzilla component mapping that OSBS does)
 - tag changes, as when an upstream `latest` could become a downstream `:release-3.4` or whatever

Please also understand that the "standard" that OSBS/IIB/CVP is using today is not a proposal for the `operator-sdk` ecosystem, as this one is. It does not provide a platform for the community at large. It is not ubiquitous nor is it applicable generally.

By explicitly expecting the user to provide a manifest that identifies the unique identifiers and their placeholders in the CSV we create an explicit, clear and straightforward interface for _all_ build systems. As you can (I hope) clearly see from the long list of implicit, semi-explicit, etc means by which OSBS currently supports an author making effectively the same decision, the state today is not simple. It is not interoperable. How might a user, today, configure that their CSV refers to an operand which may have one name upstream but another once it's processed by OSBS and renamed for a Bugzilla component? By using _common names_ which apply to the underlying identifier, a user may propose that their CSV relies on that common name _regardless of how the build system wants to rename it at build-time_.

If a build system would like to apply implicit heuristics for finding image pull specifications and provide their users with a pre-populated `image-references`, that build system may continue to do so. It will then incur _all_ of the complexity and cost that you are finding yourself managing today. The approach as used by the OpenShift release payload scales, is clear, explicit and transparent. A static analysis check to ensure that the user resolved all references is trivial to implement and CSV authors incur the cost once in the lifetime of their CSV.

Reply:
> You are inventing the requirement for authors to replcate it three times.

Given the image-references file you provided in your previous comment, could you also post the corresponding CSV file? Both must be taken into account because both are maintained by the authors. I'm really curious as to how `registry.example.com/image:tag` does not appear three times across both files in the case where disconnected support is needed.

Maybe it wasn't clear but the point I'm trying to highlight is the fact that manually maintaining placeholder pull specs across two different files is error prone. Tooling must be available to help authors with this. This is true regardless if you're building a bundle image locally or in a build system.

Reply:
There is no difference between placing an opaque string in three places or a template language placeholder. They hold identical error cases. CI that builds your bundle, installs your operator and verifies it will catch those trivially.